File: Python/BFS/M-1466. Reorder Routes to Make All Paths Lead to the City Zero.py
from collections import defaultdict, deque


class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        adj_city_map = defaultdict(set)
        for [from_city, to_city] in connections:
            adj_city_map[from_city].add((to_city, 1))
            adj_city_map[to_city].add((from_city, 0))

        queue = deque([0])
        visited = set([0])
        reorder = 0
        while queue:
            city = queue.popleft()
            for [adj_city, is_connected] in adj_city_map[city]:
                if adj_city not in visited:
                    reorder += is_connected
                    visited.add(adj_city)
                    queue.append(adj_city)
        return reorder


testData = [
    {
        "n": 6,
        "connections": [
            [0, 1],
            [1, 3],
            [2, 3],
            [4, 0],
            [4, 5],
        ],
        "ans": 3,
    },
    {
        "n": 5,
        "connections": [
            [1, 0],
            [1, 2],
            [3, 2],
            [3, 4],
        ],
        "ans": 2,
    },
]

for data in testData:
    print(data["ans"] == Solution.minReorder(data["n"], data["connections"]))

File: Javascript/BFS/M-1466. Reorder Routes to Make All Paths Lead to the City Zero.ts
function minReorder(n: number, connections: number[][]): number {
  const adjCityMap = new Map<number, Set<number[]>>();
  for (const [fromCity, toCity] of connections) {
    if (!adjCityMap.has(fromCity)) adjCityMap.set(fromCity, new Set());
    if (!adjCityMap.has(toCity)) adjCityMap.set(toCity, new Set());
    adjCityMap.get(fromCity)?.add([toCity, 1]);
    adjCityMap.get(toCity)?.add([fromCity, 0]);
  }

  let reorder = 0;
  const queue: number[] = [0];
  const visited: Set<number> = new Set([0]);
  while (queue.length) {
    const city = queue.shift();
    if (city === undefined) continue;
    for (const [adjCity, isConnected] of adjCityMap.get(city) || []) {
      if (!visited.has(adjCity)) {
        reorder += isConnected;
        visited.add(adjCity);
        queue.push(adjCity);
      }
    }
  }
  return reorder;
}

const testData = [
  {
    n: 6,
    connections: [
      [0, 1],
      [1, 3],
      [2, 3],
      [4, 0],
      [4, 5],
    ],
    ans: 3,
  },
  {
    n: 5,
    connections: [
      [1, 0],
      [1, 2],
      [3, 2],
      [3, 4],
    ],
    ans: 2,
  },
  {
    n: 3,
    connections: [
      [1, 0],
      [2, 0],
    ],
    ans: 0,
  },
  {
    n: 3,
    connections: [
      [1, 2],
      [2, 0],
    ],
    ans: 0,
  },
  {
    n: 5,
    connections: [
      [4, 3],
      [2, 3],
      [1, 2],
      [1, 0],
    ],
    ans: 2,
  },
];

for (const data of testData) {
  console.log(data['ans'] === minReorder(data['n'], data['connections']));
}
